### ISSUE TITLE
Clients / Producer not correctly dropped, leading to leak of memory

### ISSUE THREAD

**Issue by BlackYoup** [streamnative/pulsar-rs#127]
I'm having a memory leak in a project where I create multiple clients and a producer by client to send a single message. Once done, both of the client and producer are dropped but the memory still grows.

Here's the project I used to reproduce the issue:

Cargo.toml
```toml
[package]
name = "pulsar-rs-leak"
version = "0.1.0"
authors = ["arnaud"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
pulsar = "1.0"
tokio = "0.2"
```

main.rs:
```rust
use std::error::Error;

use pulsar::{Authentication, Pulsar, TokioExecutor};


#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    loop {
        let token = std::env::var("PULSAR_TOKEN").unwrap();
        let authentication = Authentication {
            name: "token".to_string(),
            data: token.into_bytes(),
        };

        let builder = Pulsar::builder("<pulsar>", TokioExecutor)
            .with_auth(authentication);

        let pulsar = builder.build().await.map_err(|e| {
            eprintln!("Error when connecting to pulsar: {}", e);
            e
        })?;

        let mut pubkey_producer = pulsar
            .producer()
            .with_topic("<topic>")
            .with_name("memory-leak-test")
            .build()
            .await?;

        println!("Created producer");
    }

    Ok(())
}
```

I then run `cargo run` and in another terminal, I'm watching the memory: `watch -n1 'ps -eo args,%mem | grep -E '[l]eak''`. The memory grows indefinitely. It is faster in `--release` mode.

If I drop the producer creation, the memory still seems to leak but at a slower rate. But that may only be an impression.

**Comment by Geal** [comment]
thanks for the report, this should be fixed by 67f7c30c3817e

**Comment by Geal** [comment]
closing this now, feel free to reopen if it reappears
